examples/abitest/module_0/rust/src/lib.rs

```diff
@@ -753,6 +757,65 @@ impl FrontendNode {
         Ok(())
     }
 
+    fn test_handle_clone_raw(&mut self) -> TestResult {
+        let (w, r, _) = channel_create_raw();
+
+        unsafe {
+            // Clone both handles
+            let mut w2 = 0u64;
+            let mut r2 = 0u64;
+            expect_eq!(
+                OakStatus::Ok as u32,
+                oak_abi::handle_clone(w, &mut w2 as *mut u64)
+            );
+            expect_eq!(
+                OakStatus::Ok as u32,
+                oak_abi::handle_clone(r, &mut r2 as *mut u64)
+            );
+
+            // Handles should be distinct values
+            expect!(w != w2);
+            expect!(r != r2);
+
+            // Close the original handles
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(w));
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(r));
+
+            // Check that we can close the cloned handles (and thus the channel remains open)
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(w2));
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(r2));
+
+            // Check that an invalid handle cannot be cloned
+            let bogus_handle = 99999;
+            let mut cloned_bogus_handle = 0u64;
+            expect_eq!(
+                OakStatus::ErrBadHandle as u32,
+                oak_abi::handle_clone(bogus_handle, &mut cloned_bogus_handle as *mut u64)
+            );
+
+            // A handle to a closed channel cannot be cloned
+            let (w, r, _) = channel_create_raw();
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(w));
+            expect_eq!(OakStatus::Ok as u32, oak_abi::channel_close(r));
+            let closed_handle = r;
+            let mut cloned_closed_handle = 0u64;
+            expect_eq!(
+                OakStatus::ErrBadHandle as u32,
+                oak_abi::handle_clone(closed_handle, &mut cloned_closed_handle as *mut u64)
+            );
+
+            // Invalid handle_out parameter returns error
```

Comment:
Would it be possible to restrict the `unsafe` block to just this bit of the test?

(Might even be worth splitting this into a `ChannelClone` + `ChannelCloneRaw` pair of tests)

Reply:
Aah, I was using `oak_abi::channel_close` instead of the safe `oak::channel_close`. By changing that only the calls to `oak_abi::handle_clone` are `unsafe`, I'll make the `unsafe` scope a bit smaller.

My impression was that the `..Raw` tests check that function in `oak_abi`, whereas the others check calls in `oak`. Should I reserve `ChannelClone` for tests of the form: `my_handle.clone()`?

Reply:
> My impression was that the ..Raw tests check that function in oak_abi, whereas the others check calls in oak.

Yeah, that was roughly the idea &ndash; I tend to think of `BaseTest` as exercising all the misbehaviour an Oak app can get up to with (just) the `sdk` wrappers, whereas the `BaseTestRaw` variant allows additional misbehaviour that's only possible with the raw Wasm ABI calls (mostly invalid linear addresses).  

The distinction probably hasn't been applied consistently though, so it's not a big deal if you want to leave as-is.

Reply:
Sounds good, I've renamed the test to `HandleCloneRaw`, and I think it captures the misbehaviour using raw ABI calls as described. Once we actually implement `Clone` for the handle types in the SDK, I'll add another test `HandleClone`.